### DISCUSSION TITLE
Is there a recommended structure for applications?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#9283] [Usage Questions]
I'm wondering about the best way to structure an application with sqlalchemy.

Currently I have something like:
base.py -> Base
object_1.py -> Object1(Base)
object_2.py -> Object2(Base)
app.py -> creates views from views.py
views.py -> creates handlers which use Object1/2.

The problem is that the models end up in the metadata implicitly, rather than explicitly. e.g. When running the application, it is run through app.py which then creates the handlers in view.py which imports the models because it uses them there, and only because they were imported do they become part of the metadata.

As a result, when setting up alembic, I did the obvious thing and used `from .base import Base` to use for the metadata object. Obviously, that module doesn't import any of the models, so I got an empty DB from alembic. I can instead ensure that I import app.py or something, which triggers all the other application imports to happen, so it ends up with the configured metadata, but this doesn't feel explicit.

So, I'm wondering if there's a more explicit approach, or a recommended structure to complex applications?

**Comment by lelit** [comment]
I usually structure my apps like you said, and in the module's `__init__.py` I explicitly import all the the classes. See [here](https://gitlab.com/metapensiero/SoL/-/blob/master/src/sol/models/__init__.py#L181) for an example.

**Reply by Dreamsorcerer** [reply]
Hmm, that looks like a circular import to me..

**Reply by lelit** [reply]
Where do you see a circularity?

**Reply by Dreamsorcerer** [reply]
Base is defined in the file you pointed to, which then imports the other models, which import Base from the previous file...

**Reply by Dreamsorcerer** [reply]
I assume the only reason it works is because your imports are in the middle of the file, after defining Base, which is not typical recommended practice.

In terms of being explicit, I was thinking maybe it could be possible to do something like `class Object1(DeclarativeBase)` (without a user-defined `Base`), and then elsewhere in the code, you could do `Metadata(Object1, Object2)`, or similar. So, the models in the metadata are included explicitly, not dependent on what you happen to import at any given location.

**Reply by CaselIT** [reply]
No, using declarative base directly does not work. You need to create an user defined base class

**Reply by Dreamsorcerer** [reply]
Yeah, I know, I just meant as a vague possibility, rather than something that works today.

**Reply by CaselIT** [reply]
Btw I also am not sure why it is a problem.

I usually use a package for the models. There I have a module for each model plus one for the base/configuration. The init of the package exports all the single models. There is no circular import

**Reply by Dreamsorcerer** [reply]
Yeah, if Base is not created in `__init__.py`, then that seems reasonable. You potentially lose some flexibility with namespacing (e.g. you'd be expected to import `db.UserAccount` instead of `user.Account` or similar), but if that's the expected way to do it, I'll look at reorganising my projects along those lines.

**Reply by CaselIT** [reply]
I think you can import from the module package, or directly from the module, but of course the module is inside the package 

Also this is just a suggestion. Many people also use a single file for all the models, that may work too if the project is bot very big,

**Comment by lelit** [comment]
That's just a detail: in a much more complex application, where I have a database with a bunch of different [schematas](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.params.schema), I have a structure like
```
model/
├── __init__.py
├── entities
│   ├── __init__.py
│   ├── auth
│   ├── celery
│   ...
│   └── media
├── tables
│   ├── __init__.py
│   ├── auth
│   ├── celery
│   ...
│   ├── media
│   └── meta.py
└── utils.py
```

where `model/tables/__init__.py` contains

```python
from . import (
    auth,
    celery,
    ...
    media,
)
```

and `model/tables/meta.py` defines abstract tables and the common metadata instance

```python
import sqlalchemy as sa

...

Table = sa.Table

class TimeStampedTable(Table):
    """Specialized :class:`Table` subclass, carrying common fields."""

metadata = sa.MetaData()
"Container of all tables meta information."

integer_t = types.Integer()
"SA equivalent of :term:`integer_t`."

...
```

and a typical table definition, say `model/tables/celery/periodic_tasks.py`, contains

```python
import sqlalchemy as sa

from ... import translatable_string as _
from .. import meta

periodic_tasks = meta.Table(
    'periodic_tasks', meta.metadata,
    sa.Column('id', meta.integer_t,
              primary_key=True,
              nullable=False,
              info=dict(label=_('ID'),
                        hint=_('The unique ID of the task.'))),
    ...
    schema='celery')
```

Look ma, no "circular" imports!

**Comment by jenstroeger** [comment]
Just the thread I was looking for… (And perhaps just a little related: https://github.com/sqlalchemy/sqlalchemy/discussions/7906 https://github.com/sqlalchemy/sqlalchemy/discussions/9263.)

Regarding circular dependencies and one-module-per-mapping: using type hints for FKs and relationship mappers, I can quickly end up with a maze of circular dependencies if the data model gets more complex (messy).

In the past and for smaller apps I then moved all mapper classes into a single `mappers.py` module, and defined the relationships right alongside the FKs inside of the mapper classes.

However, some folks frown upon large file sizes (personally, I don’t) and resist the notion of a single module that contains all mapper classes.

There’s also the option to [late-evaluate relationship argument](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#late-evaluation-of-relationship-arguments) which I’ve not yet explored to its fullest of capabilities.

So my question — what’s the recommendation for more complex models and web-y relationships? One file, multiple, late-evaluation, … ?

Thank you!
